message view: Allow muted user's message to be rehidden.

m-e-l-u-h-a-n opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
An new option is added in message action popover of a muted user
message that allows a muted user's message to be rehidden. 
Fixes: #18452.

**Testing plan:** Manually and existing node tests.


**GIFs or screenshots:** 
![rehide](https://user-images.githubusercontent.com/63504956/118269724-b35db600-b4dc-11eb-980a-479eb34a3a07.gif)

m-e-l-u-h-a-n commented:
cc: @abhijeetbodas2001 as you opened the issue and worked on most of mute user code.

abhijeetbodas2001 commented on static/templates/actions_popover_content.hbs:
```diff
@@ -52,6 +52,14 @@
+        <a href="#" class="rehide_muted_user_message" data-message-id="{{message_id}}">
+            <i class="fa fa-minus" aria-hidden="true"></i>
+            {{t "Hide this message" }}
+        </a>
```
Would it be more clear that this is related to user muting (and not a completely different feature, like "Collapse") if it included the word "again"? Something like `Hide this message again` or `Undo reveal`?

m-e-l-u-h-a-n replied:
Both of them look good to me. May be I'll change it to `Undo reveal` as it is short and not causing confusion between the adjacent terms in the list like `Hide` and `Collapse`. 
I'll update if I get further suggestion regarding this.

alya replied:
How about this?

1. Put "Hide this message again" as the top item in the menu to make it easy to find.
2. Keep the rest of the menu the same as now for consistency.
3. Change the icon for "Hide this message again" to be an eye-slash to differentiate it from the icon for `Collapse`.

What do you think?

m-e-l-u-h-a-n replied:
Okk, I changed it as @alya suggested.

abhijeetbodas2001 commented on static/js/message_list_view.js:
```diff
@@ -1186,6 +1186,11 @@ export class MessageListView {
+    hide_revealed_message(message_id) {
+        const message_container = this.message_containers.get(message_id);
+        this._rerender_message(message_container, false, false);
+    }
```
This isn't related to this PR, but maybe it would be more readable if we change `_rerender_message` to take in a dictionary rather than three parameters like this. What do you think?

m-e-l-u-h-a-n replied:
I tam not sure if that would give any benefits also while changing this I think it would be desirable to retain the IDE support(being able to see what variable it takes by hovering over its name). For that it would have to be changed like:
```javascript
_rerender_message({message_container, message_content_edited, is_revealed})
```
so that on hovering one can know what are the properties it will use/extract. So we again have to write each of these separately. That's why I am little doubtful if this would be a useful change.
Although, I may be missing some other aspects so I think @timabbott could have better thoughts on this.

abhijeetbodas2001 replied:
My hope was to make *calling* `rerender_message` more cleaner. So, with your example, we will call `rerender_message` like the following, right?
```js
this._rerender_message({
    message_container,
    message_content_edited: false,
    is_revealed: false,
});
```

This I think is more readable than just passing boolean value, which we do now. But yeah, let's wait for more comments.

m-e-l-u-h-a-n replied:
Yeah this looks more clear about what parameters are being passed.

timabbott replied:
Yeah, it'd be well worth doing that refactor in this PR while we're doing manual testing for this feature; I had the same experience of feeling like I needed to open the file to check that those parameters do what I think they do, which would be solved with proper naming.

m-e-l-u-h-a-n replied:
Added [d20fdb7](https://github.com/zulip/zulip/pull/18497/commits/d20fdb7eec1cf3bbbe9894a25e5fe36daff898f1) for this refactoring.

abhijeetbodas2001 reviewed (commented):
Thanks for working on this! I fetched and tested it out, works great. Some questions I have are-
- Should we hide the "Collapse" button when the message is hidden/revealed? Having two options like "Hide" and "Collapse" together might be confusing to users, and I don't think there's much use of being able to collapse a hidden message anyways.
- When a hidden message is revealed by clicking, we should probably close the 3-dots menu if it is open, because it can endup in a wrong state after this. (Not having the "Hide again" option when it should have)
- Can you add tests to verify the `is_hidden` variable? `node_tests/message_list_view.js` has some setup which can be extended.

m-e-l-u-h-a-n commented:
Changes in this recent push are:
* Hide action popover(opened using three dots) before revealing the message.
* Modified `message_list_view.js` node tests to test muted user message related variables when `set_calculated_message_container_variables` is called with `is_revealed=false` which is what we basically do for re-hiding message from popover.
* Changed icons and wordings of the option in popover.

It is ready for another review.

abhijeetbodas2001 commented:
LGTM!

timabbott commented:
This looks great; we should add some mention of the fact that you can do this in the the /help/ documentation.

Also, it's worth a bit of a discussion in the chat.zulip.org thread on muting for precisely what to call the new actions menu option.

m-e-l-u-h-a-n commented:
Updated pr with following changes:
* Added [0ec8480](https://github.com/zulip/zulip/pull/18497/commits/0ec8480d92dda118c661945472d653024f4f0842) to update /help/ to mention about re-hiding muted user's messages.
* Added [0a3aeb3](https://github.com/zulip/zulip/pull/18497/commits/0a3aeb34c67aa410c1900d60482779e011a70323) for `_rerender_message` refactor.

Also [posted](https://chat.zulip.org/#narrow/stream/101-design/topic/Hiding.20user-muted.20messages/near/1183408) in chat.zulip.org regarding discussion about naming of new option.

alya commented on templates/zerver/help/mute-a-user.md:
```diff
@@ -13,7 +13,8 @@ have the following effects:
   picture, and message content, are hidden behind a **Click here to
-  reveal** banner.
+  reveal** banner which can be used to reveal such messages. It is also
+  possible to [re-hide revealed messages of muted users](/help/mute-a-user#re-hide-revealed-messages-of-muted-users).
```
Maybe something like this, to keep it shorter:

All messages sent by muted users, including the name, profile  picture, and message content, are hidden behind a **Click here to  reveal** banner.  A revealed message can later be [re-hidden](/help/mute-a-user#re-hide-revealed-messages-of-muted-users).

m-e-l-u-h-a-n replied:
Made these changes.

alya commented on templates/zerver/help/mute-a-user.md:
```diff
@@ -68,6 +69,16 @@ have the following effects:
 
 {end_tabs}
 
+### Re-hide revealed messages of muted users
```
How about, "Re-hide a message that has been revealed"?

m-e-l-u-h-a-n replied:
Posted a [comment](https://chat.zulip.org/#narrow/stream/101-design/topic/Hiding.20user-muted.20messages/near/1184731) about it on chat.zulip.org.

m-e-l-u-h-a-n replied:
Made these changes.

m-e-l-u-h-a-n commented:
This push changes name of new option in poppver to `Hide muted message again`, as discussed on chat.zulip.org. Also updated /help changes as suggested on [https://github.com/zulip/zulip/pull/18497#discussion_r635450584](https://github.com/zulip/zulip/pull/18497#discussion_r635450584).

m-e-l-u-h-a-n commented:
All feedback on this are implemented. It is ready for another look.

timabbott commented on static/js/message_list_view.js:
```diff
@@ -1163,7 +1163,7 @@ export class MessageListView {
     }
 
-    _rerender_message(message_container, message_content_edited, is_revealed) {
+    _rerender_message({message_container, message_content_edited, is_revealed}) {
```
I think we want the `message_container` to stay a positional parameter, since that feels more readable -- effectively it's "rerender this message, with these options"; I made that tweak before merging.

timabbott commented:
Merged, thanks for doing this @m-e-l-u-h-a-n!
